### PR TITLE
Snapshot Support in DuckDB

### PR DESCRIPTION
**Pull request by muralivemulapati**
Creates a snapshot at a specific snapshot id.

### CODE REVIEW

File: src/common/local_file_system.cpp [hunk 1]
```diff
@@ -1086,6 +1096,14 @@ void LocalFileSystem::MoveFile(const string &source, const string &target, optio
 	}
 }
 
+void LocalFileSystem::CopyFile(const string &source, const string &target, unique_ptr<FileHandle>& src_handle, unique_ptr<FileHandle>& dst_handle) {
+        auto source_unicode = WindowsUtil::UTF8ToUnicode(source.c_str());
```
Comment (jeremyosterhoudt) [inline src/common/local_file_system.cpp]: intent alignment

File: src/common/local_file_system.cpp [hunk 2]
```diff
@@ -1337,4 +1355,25 @@ unique_ptr<FileSystem> FileSystem::CreateLocal() {
...
 }
 
+#if defined(__DARWIN__) || defined(__APPLE__) || defined(__OpenBSD__)
+void LocalFileSystem::CopyFile(const string &source, const string &target, unique_ptr<FileHandle>& src_handle, unique_ptr<FileHandle>& dst_handle) {
+  int src_fd = src_handle->Cast<UnixFileHandle>().fd;
+  fclonefileat(src_fd, AT_FDCWD, target.c_str(), 0);
+}
+#elif _WIN32
+void LocalFileSystem::CopyFile(const string &source, const string &target, unique_ptr<FileHandle>& src_handle, unique_ptr<FileHandle>& dst_handle) {
+  throw NotImplementedException("CopyFile Unsupported");
```
Comment (jeremyosterhoudt) [inline src/common/local_file_system.cpp]: Would someone using Windows for development run into this?  If so I think we should implement it before merging.

File: src/include/duckdb/common/file_system.hpp [hunk 3]
```diff
@@ -151,7 +151,9 @@ class FileSystem {
 	//! properties
 	DUCKDB_API virtual void MoveFile(const string &source, const string &target,
 	                                 optional_ptr<FileOpener> opener = nullptr);
-	//! Check if a file exists
+        DUCKDB_API virtual void CopyFile(const string &source, const string &target, unique_ptr<FileHandle>& src_handle, unique_ptr<FileHandle>& dst_handle);
```
Comment (jeremyosterhoudt) [inline src/include/duckdb/common/file_system.hpp]: intent alignment

**Review by jeremyosterhoudt — CHANGES_REQUESTED** [review]
Overall this looks good.  There is one question and a lot of indent alignment issues (in most of the files).  I do think we should add some unit tests at least at the C-API level for the new code we've added.

File: src/common/local_file_system.cpp (hunk 2 above)
Comment (muralivemulapati) [reply src/common/local_file_system.cpp]: Windows version implemented at line 1099. Removed this code.

File: src/common/local_file_system.cpp (hunk 1 above)
Comment (muralivemulapati) [reply src/common/local_file_system.cpp]: fixed

File: src/include/duckdb/common/file_system.hpp (hunk 3 above)
Comment (muralivemulapati) [reply src/include/duckdb/common/file_system.hpp]: fixed

**Review by jeremyosterhoudt — APPROVED** [review]

**Pull request merged**